src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+   activateBuses(mEffectComponent.get(), Vst::kAudio, Vst::kOutput);
+}
+
+VST3Effect::VST3Effect(const VST3Effect& other)
+   : mModule(other.mModule), mEffectClassInfo(other.mEffectClassInfo)
+{
+   Initialize();
+   mSetup = other.mSetup;
+}
+
+VST3Effect::~VST3Effect()
+{
+   CloseUI();
+   if(mEditController)
+      mEditController->terminate();
+   if(mEffectComponent)
+      mEffectComponent->terminate();
+}
+
+VST3Effect::VST3Effect(
+   std::shared_ptr<VST3::Hosting::Module> module, 
+   VST3::Hosting::ClassInfo effectClassInfo)
+   : mModule(std::move(module)), mEffectClassInfo(std::move(effectClassInfo))
+{
+   Initialize();
+}
+
+
+PluginPath VST3Effect::GetPath()
+{
+   return VST3Utils::MakePluginPathString( { mModule->getPath() }, mEffectClassInfo.ID().toString());
+}
+
+ComponentInterfaceSymbol VST3Effect::GetSymbol()
+{
+   return wxString { mEffectClassInfo.name() };
+}
+
+VendorSymbol VST3Effect::GetVendor()
+{
+   return wxString { mEffectClassInfo.vendor() };
+}
+
+wxString VST3Effect::GetVersion()
+{
+   return mEffectClassInfo.version();
+}
+
+TranslatableString VST3Effect::GetDescription()
+{
+   //i18n-hint VST3 effect description string
+   return XO("SubCategories: %s").Format( mEffectClassInfo.subCategoriesString() );
+}
+
+EffectType VST3Effect::GetType()
+{
+   using namespace Steinberg::Vst::PlugType;
+   const auto& cats = mEffectClassInfo.subCategories();
+   if(std::find(cats.begin(), cats.end(), kFxGenerator) != cats.end())
+      return EffectTypeGenerate;
+   if(std::find(cats.begin(), cats.end(), kAnalyzer) != cats.end())
+      return EffectTypeAnalyze;
+   
+   return EffectTypeProcess;
+}
+
+EffectFamilySymbol VST3Effect::GetFamily()
+{
+   return VST3EffectsModule::PluginType;
+}
+
+bool VST3Effect::IsInteractive()
+{
+   return mEditController != nullptr;
+}
+
+bool VST3Effect::IsDefault()
+{
+   return false;
+}
+
+bool VST3Effect::IsLegacy()
+{
+   return false;
+}
+
+bool VST3Effect::SupportsRealtime()
+{
+   return true;
+}
+
+bool VST3Effect::SupportsAutomation()
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
```

Comment:
This writes a number as the key into CommandParameters.  But ParameterInfo also has short and long titles.  Maybe the title should be written, or (if that is not certain to be unique among parameters) a concatenation of a title and the id.

If you use a VST3 effect in a macro and edit the settings, that key is stored in the macro text file, and will be visible to anyone who tries to read it.

Maybe an advanced user might try to edit the text of a macro file directly, and mnemonic names for the parameters, instead of simple numbers, would be helpful.

Of course SetAutomationParameters must be correspondingly changed too.

Reply:
According to VST documentation, it's not prohibited to change parameter names. It's not difficult to append suffix if parameter has a title to a parameter key, like `0_Vol`. Then, to read it from a CommandParameters we would need to extract id number and discard suffix.